### PR TITLE
Formspec version

### PR DESCRIPTION
**Pull request by SmallJoker**
Target is to provide a sane way to introduce new element parameters without that the client skips drawing them. Backwards compatibility can be assumed for new versions, although it won't display exactly as intended.

Compatibility layer:

* New Lua API feature to check whether `formspec_version[]` may be used
* `get_player_information().formspec_version` for the maximal supported version

## To do

This PR is  Ready for Review.

## How to test
L7-21: comment and uncomment as you like. Compare what should be the same.
```Lua
core.register_on_joinplayer(function(player_)
	local inv = player_:get_inventory()
	inv:set_list("hand", {"default:cobble"})

	core.after(5, function(player)
	core.show_formspec(player:get_player_name(), "lab:testy",
		--[[ VERSION 1 PLUS REAL COORDINATES ]]
		--"size[10,8]"..
		--"no_prepend[]"..
		--"real_coordinates[true]"..

		--[[ VERSION 2 (REAL COORDINATES) ]]
		--"formspec_version[2]"..
		--"size[10,8]"..
		--"no_prepend[]"..

		--[[ FUTURE VERSION, MORE PARAMS ]]
		"formspec_version[3]"..
		"size[10,8]"..
		"no_prepend[]"..
		"list[current_player;main;5,6;3,2;;more;args;baa_foo;bar]"..


		"background[0,0;10,8;lab_9slice.png;false;20]" ..
		--"style_type[field;noclip=true;border=false;textcolor=#FF0]"..
		"dropdown[5.2,5;2;category;a,b,c,d,e,f,g,h;1]" ..
		"list[current_player;main;5,6;3,2;]"..
		-- Book edit
		"field[0.5,1;3.5,0;title;Title:;" ..
			core.formspec_escape("Taking the hobbits to Isengard") .. "]" ..
		"textarea[0.5,1.5;3.5,4;text;Contents:;" ..
			core.formspec_escape("The hobbits the hobbits\nto Isengard") .. "]" ..
		-- Empty book
		"field[4.5,1;3.5,0;title2;"..minetest.colorize("#0F0", "Title:") .. ";]" ..
		"label[4,3;"..minetest.colorize("#0F0", "ffetre") .. "dasds]" ..
		"textarea[0.5,5.5;3.5,4;text2;Contents:;]" ..
		-- Server description (pre e6e5fa3, 0.4.16 stable)
		"textarea[4,2.3;4.23,2.9;;" ..
minetest.colorize("#F00", [[Madman drummer bummers
Indians in the summer
With a teenage diplomat
In the dumps with the mumps
As the adolescent pumps]] .. ";]") ..
		-- Readonly textarea with title
		"textarea[4.6,5.85;3.7,1.5;;This is a fancy title" ..
		";Never gonna give you up, never gonna let you down]"
	)
	end, player_)
end)
```

### CODE REVIEW

File: src/gui/guiFormSpecMenu.h [hunk 1]
```diff
@@ -472,7 +472,7 @@ class GUIFormSpecMenu : public GUIModalMenu
 private:
 	IFormSource        *m_form_src;
 	TextDest           *m_text_dst;
-	u32                 m_formspec_version = 0;
```
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.h]: This change does not matter, because the version was never increased. 0 behaves the same as 1.

**Comment by Wuzzy2** [comment]
Given the new field `formspec_version` will probably become very important soon, I suggest to make the name shorter. I suggest `version`, maybe even just `v` because it will probably be used (maybe even required, practically speaking) in all future formspecs. The `formspec_` prefix is unneccessary IMO, every formspec element has something to do with formspecs. :-)

Questions:

* Do you have a defined process on what conditions to bump the version? Is it written down?
* Are the any “safe” changes to formspec syntax, that don't require version bump? Is is written down?
* Do you expect any other major changes in the future that would require a version bump? Can they be avoided?
* When you bump version to 3 (or 4, 5, 6, etc.), will you retain the documentation of the old version? It might be required to do so for migration purposes … :-/

**Comment by Wuzzy2** [comment]
Another question: Given that `real_coordinates` was never in a stable release, and `formspec_version` was also never in a stable release, and `formspec_version[2]` triggers real coordinates, would it make sense to just drop `real_coordinates` to keep things simple?

**Comment by v-rob** [comment]
I never really liked the solution of `real_coordinates` changing the form coordinates when it's placed right after `size`, `position`, `anchor`, and `no_prepend`.  `formspec_version` would be a much cleaner solution.  I agree that `version` is better than `formspec_version`.  The less bytes we have to send, the faster it is for slower connections.

**Comment by Wuzzy2** [comment]
I was more concerned about readability, and that the `formspec_` prefix feels redundant, but yeah, byte count is also an argument. :-)

**Comment by DS-Minetest** [comment]
Renaming `formspec_version` to `version` would probably cause compatibility problems (see also [here](https://github.com/minetest/minetest/blob/720aedb46740ba203b4a0f3e049e368abc4932b3/src/gui/guiFormSpecMenu.cpp#L2004)).

**Comment by Wuzzy2** [comment]
Why?

**Comment by rubenwardy** [comment]
The engine already had partial support for formspec versions. Renaming the tag would prevent this

You could do a string replace if the first characters are "version[" to work around this

**Comment by SmallJoker** [comment]
> would it make sense to just drop real_coordinates to keep things simple?

I also thought about that. Yes, unless someone would want to embed code with the older coordinates - in that case they could prepend `real_coordinates[false]` to inject that particular code. Not needed for me, and barely for anyone else. Removing it would be OK, but not in this PR.

> and that the formspec_ prefix feels redundant,

As @DS-Minetest already pointed out, older clients (namely 5.0.x) would assume version 1 for any formspec, and even throw warnings because there's an unknown element before `size[]`.

**Comment by SmallJoker** [comment]
9acd36b

**Pull request closed without merging**